### DISCUSSION TITLE
Rate Limit handling

### DISCUSSION THREAD

**Discussion by marcogruhl** [Xwilarg/TwitterSharp#11] [Ideas]
Hi,
this wrapper is the best i found for my twitter project. I played around and hit constantly rate limits. Sadly, this information is not provided.
I look at the code and have **two ideas** to implement this based on [Twitter API Documentation](https://developer.twitter.com/en/docs/twitter-api/rate-limits#headers-and-codes)

the twitter has to use `GetAsync` instead of `GetStringAsync`

### 1. Extend the Response objects

`public class StreamInfo : XRate, IEquatable<StreamInfo>`

the XRate Object would look like:

```
public class XRate
{
    public int Limit { get; set; }
    public int Remaining { get; set; }
    public int Reset { get; set; }
}
```
and than use the `ParseArrayData` Method like this:

! VERY UNOPTIMIZED !

```
private T[] ParseArrayData<T>(string json, HttpResponseHeaders headers = null) where T : XRate
{
    var answer = JsonSerializer.Deserialize<Answer<T[]>>(json, _jsonOptions);
    if (answer.Detail != null)
    {
        throw new TwitterException(answer.Detail);
    }
    if (answer.Data == null)
    {
        return Array.Empty<T>();
    }
    InternalIncludesParse(answer);
    
    if(headers != null)
    {
        if (headers.TryGetValues("x-rate-limit-limit", out var limit))
        {
            foreach (var data in answer.Data)
            {
                data.Limit = Convert.ToInt32(limit.FirstOrDefault());
            }
        }
        if (headers.TryGetValues("x-rate-limit-remaining", out var remaining))
        {
            foreach (var data in answer.Data)
            {
                data.Remaining = Convert.ToInt32(remaining.FirstOrDefault());
            }
        }
        if (headers.TryGetValues("x-rate-limit-reset", out var reset))
        {
            foreach (var data in answer.Data)
            {
                data.Reset = Convert.ToInt32(reset.FirstOrDefault());
            }
        }
    }

    return answer.Data;
}
```
### 2. Provide via EventHandler

TweetClient.cs:
```
public event EventHandler<XRate> XRateChanged;

public async Task<StreamInfo[]> GetInfoTweetStreamAsync()
{
    // var str = await _httpClient.GetStringAsync(_baseUrl + "tweets/search/stream/rules");
    var str = await _httpClient.GetAsync(_baseUrl + "tweets/search/stream/rules");

    XRateChanged.Invoke(this, new XRate(nameof(StreamInfo), str.Headers));

    return ParseArrayData<StreamInfo>(await str.Content.ReadAsStringAsync(), str.Headers);
}
```
XRate Class:
```
public class XRate
{
    public string Endpoint { get; set; }
    public int Limit { get; set; }
    public int Remaining { get; set; }
    public int Reset { get; set; }

    public XRate()
    {
        
    }

    public XRate(string endpoint, HttpResponseHeaders headers)
    {
        Endpoint = endpoint;

        if (headers != null)
        {
            if (headers.TryGetValues("x-rate-limit-limit", out var limit))
                Limit = Convert.ToInt32(limit.FirstOrDefault());
            
            if (headers.TryGetValues("x-rate-limit-remaining", out var remaining))
                Remaining = Convert.ToInt32(remaining.FirstOrDefault());

            if (headers.TryGetValues("x-rate-limit-reset", out var reset))
                Reset = Convert.ToInt32(reset.FirstOrDefault());
        }
    }
}
```
I will implement this if its fitting to your idea of this project

**Comment by Xwilarg** [comment]
I like the idea of extending the Response object but maybe headers shouldn't be parsed by default on every request?

What do you think of having a BaseObject class like that:
```cs
public class BaseObject
{
    internal BaseObject(HttpResponseHeaders headers)
        => _headers = headers;

    public XRate GetRateLimits()
    {
        // Implement header parsing as shown above
    }

    private HttpResponseHeaders _headers;
}
```

And every result object would inherit it so GetRateLimits() would be accessible from every response?

**Reply by marcogruhl** [reply]
The last days i tried a lot of ways to implement this. I will give an as short as possible recap of this and my conclusion for now.

TLDR; 7. is the best solution i found so far.

### 1. BaseObject

- Its getting complicated, when deserialisation has to call the non-parameterless constructor, especially when we deal with T, T[], or object.
- Enrich ever partial response object is not the exact place for the rate limit. It should be at the request level.

### 2. BaseObject without non-parameterless constructor
```C#
public void AddHeaders(HttpResponseHeaders headers)
            => _headers = headers;
```
- Enrich ever partial response object is not the exact place for the rate limit. It should be at the request level.
- the implementation must be very specifict for every response e.G.:
```C#
public async Task<StreamInfo[]> GetInfoTweetStreamAsync()
{
    var res = await _httpClient.GetAsync(_baseUrl + "tweets/search/stream/rules");
    var obj = ParseArrayData<StreamInfo>(await res.Content.ReadAsStringAsync());

    foreach (var streamInfo in obj)
        streamInfo.AddHeaders(res.Headers);

    return obj;
}
```
### 3. Provide via EventHandler (former 2.)

- Its not directly connected to the request

### 5. Response wrapping Object
Create a Response object for every Request
```C#
public class StreamInfoResponse : BaseResponse, IEnumerable<StreamInfo>
{
    public IEnumerable<StreamInfo> StreamInfos { get; set; }

    public StreamInfoResponse(IEnumerable<StreamInfo> streamInfos, HttpResponseHeaders headers)
    {
        StreamInfos = streamInfos;
        _headers = headers;
    }

    public IEnumerator<StreamInfo> GetEnumerator()
    {
        foreach (var streamInfo in StreamInfos)
        {
            yield return streamInfo;
        }
    }

    IEnumerator IEnumerable.GetEnumerator()
    {
        return GetEnumerator();
    }
}
```
- Eventually braeking changes, although you can use it as enumarable
- Not very generic

### 6. Return generic Tuple
```C#
public async Task<(StreamInfo[] Data, RateLimit RateLimit)> GetInfoTweetStreamAsync()
{
    var res = await _httpClient.GetAsync(_baseUrl + "tweets/search/stream/rules");
    return (ParseArrayData<StreamInfo>(await res.Content.ReadAsStringAsync()), new RateLimit(res.Headers));
}
```
- breaking changes
- may not work for every request

### 7. Action return RateLimit <- favored
```C#
public async Task<StreamInfo[]> GetInfoTweetStreamAsync(Action<RateLimit> getRateLimit = null)
{
    var res = await _httpClient.GetAsync(_baseUrl + "tweets/search/stream/rules");

    if(getRateLimit != null)
        getRateLimit.Invoke(new RateLimit(res.Headers));

    return ParseArrayData<StreamInfo>(await res.Content.ReadAsStringAsync());
}
```
- no breaking changes
- optional for caller and not parsed on every request
- relatively generic to implement

Have you any further comments to this?

**Reply by Xwilarg** [reply]
I'm just having a quick look before going to sleep, but wouldnt 7 means adding a new parameter for all the public methods? Since you might want to get the rate limit not only for the GetInfoTweetStream

**Reply by marcogruhl** [reply]
Yes, thats the downside of this approach.
i implemented it in this branch [RateLimit](https://github.com/marcogruhl/TwitterSharp/tree/RateLimit)
With adding it to all methods it kind of "conflicts" with the `params`, which had to be the last parameter. So there are also some minor breaking changes.

**Reply by Xwilarg** [reply]
To go back on what you said before, what exactly is the issue with 1?

**Reply by marcogruhl** [reply]
As far as i see all Objects were created in the `ParseData` and `ParseArrayData` functions with `JsonSerializer.Deserialize`. I dont know (yet) how i can inject an (external) parameter in the constructor.

**Reply by Xwilarg** [reply]
Well it's in the headers anyway so I don't think JsonSerializer will help much

**Reply by marcogruhl** [reply]
Sorry, I dont understand. At least idea 1 and 2 have problems with `DeleteTweetStreamAsync` and its `ParseData<object>` return and `NextTweetStreamAsync` which returns not until the first tweet. Maybe idea 3 is the least breaking version.

**Reply by Xwilarg** [reply]
Uh yeah I forgot about that  
I feel like event would not be the best solution but at the same time I don't have any other idea

**Reply by marcogruhl** [reply]
I implemented the EventHandler logic wich has no breaking changes: [EventHandler branch](https://github.com/marcogruhl/TwitterSharp/tree/RateLimitEventHandler)
Can you look at it? Should i create a pull request? Do i have to extend the tests?
Sorry for those questions, it would be my first public pull request.

**Reply by Xwilarg** [reply]
Yup you can create a PR, that will also allow the current tests to run and see if everything still works well  
It would indeed also be nice if you could add some tests to them, maybe like, do a request, get current rate limit and see if it changes when you do another one?

**Comment by marcogruhl** [comment]
Created the pull request #14 
I also extended the test `TestStreamProcess()` because there are already a method which is called multiple times, so i can check the remaining rate limit decreasing like you suggested.
